experimental/gittuf/root.go

```diff
@@ -776,6 +782,166 @@ func (r *Repository) RemovePropagationDirective(ctx context.Context, signer ssli
+		}
+	}
+
+	options := &trustpolicyopts.Options{}
+	for _, fn := range opts {
+		fn(options)
+	}
+
+	rootKeyID, err := signer.KeyID()
+	if err != nil {
+		return err
+	}
+
+	slog.Debug("Loading current policy...")
+	state, err := policy.LoadCurrentState(ctx, r.r, policy.PolicyStagingRef)
+	if err != nil {
+		return err
+	}
+
+	if hookName == "" {
+		hookName = filepath.Base(filePath)
+	}
+
+	rootMetadata, err := r.loadRootMetadata(state, rootKeyID)
+	if err != nil {
+		return err
+	}
+
+	slog.Debug("Checking if hook with specified name already exists...")
+	found := false
+	for _, hook := range state.Hooks[stage] {
+		if hook.ID() == hookName {
+			found = true
+		}
+	}
+
+	if found {
+		return tuf.ErrDuplicatedHookName
+	}
+
+	slog.Debug("Reading hook from filesystem...")
+	hookFile, err := os.Open(filePath)
+	if err != nil {
+		return err
+	}
+	defer hookFile.Close() // nolint:errcheck
+
+	hookFileContents, err := io.ReadAll(hookFile)
+	if err != nil {
+		return err
+	}
+
+	var hashes = make(map[string]string, 2)
+	blobID, err := r.r.WriteBlob(hookFileContents)
+	if err != nil {
+		return err
+	}
+	//TODO: hash agility
+	hashes["sha1"] = blobID.String()
+
+	sha256Hash := sha256.New()
+	sha256Hash.Write(hookFileContents)
+	hashes["sha256"] = hex.EncodeToString(sha256Hash.Sum(nil))
+
+	slog.Debug("Adding hook to rule file...")
+	if err := rootMetadata.AddHook(stage, hookName, principalIDs, hashes, environment, modules); err != nil {
+		return err
+	}
+
+	hook := &tufv01.Hook{
+		Name:         hookName,
+		PrincipalIDs: set.NewSetFromItems(principalIDs...),
+		Hashes:       hashes,
+		Environment:  environment,
+		Modules:      modules,
+	}
+
+	state.Hooks[stage] = append(state.Hooks[stage], hook)
+
+	commitMessage := fmt.Sprintf("Add hook '%s' to root metadata", hookName)
+
+	slog.Debug("Committing policy...")
+	return r.updateRootMetadata(ctx, state, signer, rootMetadata, commitMessage, options.CreateRSLEntry, signCommit)
+}
+
+// RemoveHook defines the workflow for removing a hook defined in gittuf policy.
+func (r *Repository) RemoveHook(ctx context.Context, signer sslibdsse.SignerVerifier, stage tuf.HookStage, hookName string, signCommit bool, opts ...trustpolicyopts.Option) error {
+	if !dev.InDevMode() {
+		return dev.ErrNotInDevMode
+	}
+
+	if signCommit {
+		slog.Debug("Checking if Git signing is configured...")
+		err := r.r.CanSign()
+		if err != nil {
+			return err
+		}
+	}
+
+	options := &trustpolicyopts.Options{}
+	for _, fn := range opts {
+		fn(options)
+	}
+
+	rootKeyID, err := signer.KeyID()
+	if err != nil {
+		return err
+	}
+
+	slog.Debug("Loading current policy...")
+	state, err := policy.LoadCurrentState(ctx, r.r, policy.PolicyStagingRef)
+	if err != nil {
+		return err
+	}
+
+	rootMetadata, err := r.loadRootMetadata(state, rootKeyID)
+	if err != nil {
+		return err
+	}
+
+	slog.Debug("Checking if hook with specified name exists...")
+	hooks := []tuf.Hook{}
+	found := false
+	for _, hook := range state.Hooks[stage] {
```

Comment:
this can panic if invoked on a stage that doesn't exist, unless you're serializing an empty list for each stage?

Reply:
fwiw, we shouldn't be serializing empty lists.